src/Cli/dotnet/Commands/Run/Api/RunApiCommand.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.CommandLine;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.DotNet.Cli.Commands.Run.Api;
+
+/// <summary>
+/// Takes JSON from stdin lines, produces JSON on stdout lines, doesn't perform any changes.
+/// Can be used by IDEs to see the project file behind a file-based program.
+/// </summary>
+internal sealed class RunApiCommand(ParseResult parseResult) : CommandBase(parseResult)
+{
+    public override int Execute()
+    {
+        for (string line; (line = Console.ReadLine()) != null;)
+        {
+            if (string.IsNullOrWhiteSpace(line))
+            {
+                continue;
+            }
+
+            try
+            {
+                RunApiInput input = JsonSerializer.Deserialize(line, RunFileApiJsonSerializerContext.Default.RunApiInput);
+                RunApiOutput output = input.Execute();
+                Respond(output);
+            }
+            catch (Exception ex)
+            {
+                Respond(new RunApiOutput.Error { Message = ex.Message, Details = ex.ToString() });
+            }
+        }
+
+        return 0;
+
+        static void Respond(RunApiOutput message)
+        {
+            string json = JsonSerializer.Serialize(message, RunFileApiJsonSerializerContext.Default.RunApiOutput);
+            Console.WriteLine(json);
+        }
+    }
+}
+
+[JsonDerivedType(typeof(GetProject), nameof(GetProject))]
+internal abstract class RunApiInput
+{
+    private RunApiInput() { }
+
+    public abstract RunApiOutput Execute();
+
+    public sealed class GetProject : RunApiInput
+    {
+        public string? ArtifactsPath { get; init; }
+        public required string EntryPointFileFullPath { get; init; }
```

Comment:
This represents the primary bit of information that IDE must provide, right? This means that IDE must come up with a solution for searching, from a loose file that was opened, to find a "nearest relevant entry point", cracking the files as we go, and pass that in to dotnet cli. I think that's something we can do, just making sure of the expectation.

Reply:
This API is for single-file currently, so it should be simple - just pass the current file as the EntryPointFileFullPath.

For multi-file scenarios, I agree you would need to do the search you describe. The API will then also take the list of files to exclude (i.e., the other entry points). The implementation can automatically discover the other `.cs` files to include (which it needs to read for directives) based on the entry point or we can pass them to the API if that would be better for perf (since the IDE might have already discovered them via its more efficient file watchers).

Reply:
Btw, it might be useful if the IDE checks SDK's version to determine whether run-api is present and whether it supports single-file or multi-file scenarios - so the IDE can give better error messages.

Reply:
Maybe a version stamp in requests/responses would give us enough expressiveness here? That would be a bit like [SymbolKey.FormatVersion](https://github.com/dotnet/roslyn/blob/54481df62931af869d5169efdb0a009dfda0814c/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.cs#L119).